### PR TITLE
Feature bokeh dof

### PR DESCRIPTION
**Pull request by DBouma**
* Improved Depth of field to use HDR
* Code cleanup on depth of field
* Changed RTV formats from 32 bit to 16 bit for reduced memory usage and increased stability
* Fixed deferred / hybrid tasks to sample correct pixels
* Changed Depth of Field tasks to use proper SRV/UAV binding
* Fixed several minor problems related to depth of field

### CODE REVIEW

File: resources/shaders/accumulation.hlsl [hunk 1]
```diff
@@ -16,8 +16,12 @@ void main(uint3 DTid : SV_DispatchThreadID)
 {
 	float2 resolution;
 	input.GetDimensions(resolution.x, resolution.y);
-	
-	float3 color = input[DTid.xy].rgb / (frame_idx + 1);
+
```
Comment (VZout) [inline resources/shaders/accumulation.hlsl]: incorrect merge?

File: resources/shaders/deferred_composition.hlsl [hunk 2]
```diff
@@ -77,13 +81,13 @@ void main_cs(int3 dispatch_thread_id : SV_DispatchThreadID)
 		// Get irradiance
 		float3 irradiance = lerp(
 			irradiance_map.SampleLevel(linear_sampler, flipped_N, 0).xyz,
-			screen_space_irradiance[screen_coord].xyz,
+			screen_space_irradiance.SampleLevel(linear_sampler, uv, 0).xyz,
```
Comment (VZout) [inline resources/shaders/deferred_composition.hlsl]: Shouldn't be linear.

File: resources/shaders/deferred_composition.hlsl [hunk 3]
```diff
@@ -63,10 +67,10 @@ void main_cs(int3 dispatch_thread_id : SV_DispatchThreadID)
...
 	{
 		// GBuffer contents
-		float3 albedo = gbuffer_albedo_roughness[screen_coord].xyz;
-		const float roughness = gbuffer_albedo_roughness[screen_coord].w;
-		float3 normal = gbuffer_normal_metallic[screen_coord].xyz;
-		const float metallic = gbuffer_normal_metallic[screen_coord].w;
+		float3 albedo = gbuffer_albedo_roughness.SampleLevel(point_sampler, uv, 0).xyz;
+		const float roughness = gbuffer_albedo_roughness.SampleLevel(point_sampler, uv, 0).w;
+		float3 normal = gbuffer_normal_metallic.SampleLevel(point_sampler, uv, 0).xyz;
+		const float metallic = gbuffer_normal_metallic.SampleLevel(linear_sampler, uv, 0).w;
```
Comment (VZout) [inline resources/shaders/deferred_composition.hlsl]: Shouldn't be linear.

File: resources/shaders/deferred_composition.hlsl [hunk 4]
```diff
@@ -103,7 +107,7 @@ void main_cs(int3 dispatch_thread_id : SV_DispatchThreadID)
 			// Sample from environment if it IS NOT hybrid rendering
 			sampled_environment_map,
 			// Reflection buffer if it IS hybrid rendering
-			buffer_refl_shadow[screen_coord].xyz,	
+			buffer_refl_shadow.SampleLevel(linear_sampler, uv, 0).xyz,
```
Comment (VZout) [inline resources/shaders/deferred_composition.hlsl]: Shouldn't be linear.

File: resources/shaders/deferred_composition.hlsl [hunk 5]
```diff
@@ -92,7 +96,7 @@ void main_cs(int3 dispatch_thread_id : SV_DispatchThreadID)
 			// Do deferred shadow (fully lit for now)
 			1.0,
 			// Shadow buffer if its hybrid rendering
-			buffer_refl_shadow[screen_coord].a,
+			buffer_refl_shadow.SampleLevel(linear_sampler, uv, 0).a,
```
Comment (VZout) [inline resources/shaders/deferred_composition.hlsl]: Shouldn't be linear.

File: resources/shaders/dof_bokeh.hlsl [hunk 6]
```diff
@@ -25,23 +25,23 @@ void main_cs(int3 dispatch_thread_id : SV_DispatchThreadID)
...
+	float2 screen_coord = int2(dispatch_thread_id.x, dispatch_thread_id.y) + 0.5f;
 	float2 texel_size = 1.0f / screen_size;
 
-	float2 uv = (screen_coord + 0.5f) / screen_size;
+	float2 uv = screen_coord / screen_size;
 
 	const uint NUMSAMPLES = NUMDOFSAMPLES * NUMDOFSAMPLES;
 	const float MAXKERNELSIZE = MAXBOKEHSIZE * 0.5f;
-	const float SHAPECURVE = 2.0f;
+	const float SHAPECURVE = 4.0f;
```
Comment (VZout) [inline resources/shaders/dof_bokeh.hlsl]: stop using all caps variables.

File: resources/shaders/rt_hybrid.hlsl [hunk 7]
```diff
@@ -111,8 +111,8 @@ float3 TraceReflectionRay(float3 origin, float3 norm, float3 direction, uint ran
 		Scene,
 		RAY_FLAG_NONE,
 		//RAY_FLAG_ACCEPT_FIRST_HIT_AND_END_SEARCH
-		// RAY_FLAG_CULL_BACK_FACING_TRIANGLES,
-		~0, // InstanceInclusionMask
+		//RAY_FLAG_CULL_BACK_FACING_TRIANGLES,
+		0xFF, // InstanceInclusionMask
```
Comment (VZout) [inline resources/shaders/rt_hybrid.hlsl]: Why this change?

File: src/scene_graph/camera_node.hpp [hunk 8]
```diff
@@ -89,6 +89,7 @@ namespace wr
 		float m_shape_amt;
 		int m_aperture_blades;
 		bool m_enable_dof = false;
+		bool m_enable_bloom = false;
```
Comment (VZout) [inline src/scene_graph/camera_node.hpp]: Bloom shouldn't be part of the camera. use the settings for render task system if you want settings. (See hbao+ task)

File: resources/shaders/rt_hybrid.hlsl (hunk 7 above)
Comment (DBouma) [reply resources/shaders/rt_hybrid.hlsl]: During the investigation of the incorrect pixel sampling I came across a very recent DXR tutorial, it was very explicit in everything that it was doing, I took it over in the hope of it making a difference however ~0 and 0xFF are the same thing. I guess it's personal preference. It slipped through a commit it seems.

File: src/scene_graph/camera_node.hpp (hunk 8 above)
Comment (DBouma) [reply src/scene_graph/camera_node.hpp]: I will resolve this in the bloom pull request as bloom needs a rework anyways.

File: resources/shaders/accumulation.hlsl (hunk 1 above)
Comment (DBouma) [reply resources/shaders/accumulation.hlsl]: This was the latest from master, I changed it to use SampleLevel instead.
